v2 sender unable to decode corrupted filename on ftp connection

## What happened

(Only confirmed to be affecting V2.24.08. We also know it doesn't affect V2.21.04)*

*UPDATE: it actually DOES affect v2.21.04... but only when running on python >= 3.9 so ubuntu 22.04.   the old version is safe only because it is running on python 3.6 on ubuntu 18.04.


A delivery failed when a sender tried to transfer a file.

```
2024-08-26 22:46:32,496 [ERROR] Delivery failed ./20240826T2236Z_MSC_Radar-Composite_MMHR_1km.tif
2024-08-26 22:46:32,497 [WARNING] sending again, attempt 2
2024-08-26 22:46:35,105 [ERROR] Unable to connect to the-destination (user:ec.gc.ca)
2024-08-26 22:46:35,105 [WARNING] sending again, attempt 3
2024-08-26 22:46:37,693 [ERROR] Unable to connect tothe-destination (user:ec.gc.ca)
2024-08-26 22:46:37,694 [INFO] appended to retry list file 20240826224343.792709827 http://ddsr-cmc-ops01.cmc.ec.gc.ca/ /20240826/MSC-RADAR/unique/GEOTIFF_6MIN/COMPOSITE/MMHR/20240826T2236Z_MSC_Radar-Composite_MMHR_1km.tif
```

On servers hosting V2.21.04, file transferring continued without any problem. However, on servers hosting V2.24.08 of v2, all following file transfers got appended to the retry queue.

```
for S in $(ls -d ddsr-cmc-ops0{1..8}*); do echo $S: && grep 'Unable to connect' $S/local/home/sarra/.cache/sarra/log/sr_sender_my-config*.log.2024-08-27 | wc -l ; done
ddsr-cmc-ops01: <-  V2.24.08
5080
ddsr-cmc-ops02: <-  V2.24.08
4173
ddsr-cmc-ops03: <-  V2.24.08
5581
ddsr-cmc-ops04: <-  V2.24.08
4864
ddsr-cmc-ops05: <-  V2.24.08
5299
ddsr-cmc-ops06: <-  V2.24.08
5187
ddsr-cmc-ops07: <-  V2.21.04
0
ddsr-cmc-ops08: <-  V2.21.04
0
```

The error message when debug is on

```
[DEBUG] sr_ftp nlst: ['20240826T2236Z_MSC_Radar-Composite_MMHR_1km.tif']
[DEBUG] sr_ftp ls_file_index
[ERROR] Unable to connect to the-destination (user: username)
[DEBUG] Exception details:
Traceback (most recent call last):
File "/usr/lib/python3/dist-packages/sarra/sr_ftp.py", line 220, in connect
else: self.init_file_index()
File "/usr/lib/python3/dist-packages/sarra/sr_ftp.py", line 312, in init_file_index
self.ftp.retrlines('LIST', self.ls_file_index )
File "/usr/lib/python3.10/ftplib.py", line 465, in retrlines
line = fp.readline(self.maxline + 1)
File "/usr/lib/python3.10/codecs.py", line 322, in decode
(result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb7 in position 44: invalid start byte
```

Position 44 refers to the `.` character in `20240826T2236Z_MSC_Radar-Composite_MMHR_1km.tif`

## Corrective action

To prevent file transfers from further going to the retry queue with V2.24.08, you can remove the corrupted filename on the remote server and the file transferring should resume.

UPDATE : Removing the file did make the sender stop failing on every file transfer. However, new transfer attempts will fail whenever `nlst` returns a file with `.` in its filename.

I'm still not sure why this is a problem on v2.24.08

## Permanent solution

Move the configuration to sr3

--- a/docs/source/Reference/sr3_options.7.rst
+++ b/docs/source/Reference/sr3_options.7.rst
@@ -881,6 +881,14 @@ NOTE::
   **KNOWN LIMITATION**: When *force_polling* is set, the *sleep* setting should be
   at least 5 seconds. It is not currently clear why.
 
+ftpFilenameEncoding <str> (default: utf-8)
+------------------------------------------
+
+When connecting to a FTP file server, the charset used to encode file names (as returned by the ls command, for example)
+can be different. This option provides ability to override the default. **latin-1** is sometimes needed to
+connect to older servers. (option fed to python ftplib, see documentation on that library for more information.)
+
+
 header <name>=<value>
 ---------------------
 

--- a/docs/source/fr/Reference/sr3_options.7.rst
+++ b/docs/source/fr/Reference/sr3_options.7.rst
@@ -865,6 +865,14 @@ REMARQUE::
   **LIMITATION CONNUE** : Lorsque *force_polling* est défini, le paramètre *sleep* doit être
   au moins 5 secondes. À l’heure actuelle, on ne sait pas pourquoi.
 
+ftpFilenameEncoding <str> (par défaut : utf-8)
+----------------------------------------------
+
+Lors de la connexion à un serveur de fichiers FTP, le jeu de caractères utilisé pour encoder les noms de fichiers (tel que renvoyé par la commande ls, par exemple)
+peut être différent. Cette option permet de remplacer la valeur par défaut. **latin-1** est parfois nécessaire
+pour se connecter à des serveurs plus anciens. (option transmise à python ftplib, voir la documentation de cette module pour plus d'informations.)
+
+
 header <nom>=<valeur>
 ---------------------
 

--- a/sarracenia/transfer/ftp.py
+++ b/sarracenia/transfer/ftp.py
@@ -68,6 +68,8 @@ class Ftp(Transfer):
         self.o.add_option('accelFtpgetCommand', 'str',
                           '/usr/bin/ncftpget %s %d')
 
+        self.o.add_option('ftpFilenameEncoding', 'str', 'utf-8' )
+
         logger.debug("sr_ftp __init__")
         self.connected = False
         self.ftp = None
@@ -211,7 +213,7 @@ class Ftp(Transfer):
             # implicit FTPS (usually port 990)
             elif self.tls and self.implicit_ftps:
                 ftp = IMPLICIT_FTP_TLS()
-                ftp.encoding = 'utf-8'
+                ftp.encoding = self.o.ftpFilenameEncoding
                 ftp.connect(host=self.host, port=self.port, timeout=expire)
                 ftp.login(user=self.user, passwd=unquote(self.password))
                 if self.prot_p:
@@ -223,7 +225,7 @@ class Ftp(Transfer):
                                      self.user,
                                      unquote(self.password),
                                      timeout=expire)
-                ftp.encoding = 'utf-8'
+                ftp.encoding = self.o.ftpFilenameEncoding
                 if self.prot_p: ftp.prot_p()
                 # needed only if prot_p then set back to prot_c
                 #else          : ftp.prot_c()

